### DISCUSSION TITLE
Proposal: Add transparency to editor toolbars

### DISCUSSION THREAD

**Discussion by NiceAesth** [ppy/osu#24384] [Ideas] [answered]
NOTE: Something which I forgot to mention, the images are at 80% UI scaling. Better preview images are available here https://github.com/ppy/osu/discussions/24384#discussioncomment-6575732

Something which many mappers have noted is the fact that the lazer editor is much more claustrophobic. With this discussion I will attempt to address and pinpoint the cause for this.

The proposed changes are the following:

- Make the editor UI translucent and disable masking on the playfield

| Before | After |
|--------|-------|
|   ![image](https://github.com/ppy/osu/assets/8437201/a5219560-9e51-436c-9427-2168d70535c7)    |    ![image](https://github.com/ppy/osu/assets/8437201/2c4e43e1-c7be-428c-87ef-e85953b565c9)   |

Reasoning: Massively reduces visual clutter, this is a change which many mappers have requested as it is one of the most unpleasant aspects when transitioning from the stable editor.

- Reduce hit marker opacity

| Before | After |
|--------|-------|
|   ![image](https://github.com/ppy/osu/assets/8437201/d2e5c4a6-2f45-4db1-b49c-632828e27392)    |    ![image](https://github.com/ppy/osu/assets/8437201/658cb5d5-94aa-4a33-aaff-41db085d1663)   |

Reasoning: Currently, combo numbers are quite hard to read with hit markers enabled, but a lot of mappers do still prefer to have them on. Reducing their opacity will make them much less distracting while also providing the expected feedback.

- I have also reduced the height of the SummaryTimeline / BottomBar from 60 to 50 since it is currently quite large considering the information it provides.

The branch used to create the preview images is available [here](https://github.com/NiceAesth/osu/tree/editor-transparency)
I have created this discussion since I was asked to after receiving feedback from mapping Discord servers and on [Twitter](https://twitter.com/NiceAesth/status/1684641991785639936)

**Comment by OliBomby** [comment]
I quite like the proposed change. An open editor feels less claustrophobic and the free space opens your mind for new mapping ideas like an artist looking at a blank canvas.

**Comment by bdach** [comment]
Not very happy about this being raised again given https://github.com/ppy/osu/issues/12020 and https://github.com/ppy/osu/discussions/21999 and probably 5 more threads on this exist. But alas.

The major pain point in all of this for me personally is that designs all over the game have been consistently moving in the direction of removing semi-transparent elements, which makes editor stick out like a sore thumb.

I'm also not really getting the "claustrophobia" thing. I get that objects are allowed to stick out of what is technically the "playfield bounds" because objects have a non-zero size, but it feels really odd that mappers are relying on looking at gameplay elements through half-transparent unrelated UI elements which, for the purposes of looking at a map, are visual noise that has to be filtered out?

Put differently, what is with the insistence on solving this via transparency? Is there another argument for solving the issue that way other that stable familiarity? Because the screenshots in the OP do nothing for me to prove that transparency is the solution. It's a contrived example, no real map will do anything like that.

Maybe we could have something like a "distraction-free" mode that IDEs have, which hides most "irrelevant" stuff?

**Reply by NiceAesth** [reply]
The biggest reason for wanting translucent UI is mostly related to the fact that all of the UI elements are much larger on lazer. Since vertical space is very limited, the timelines clash with objects placed near the borders of the playfield, making this an issue even for hitcircles (as you have mentioned).

![image](https://github.com/ppy/osu/assets/8437201/446442ee-511e-4e0a-9746-d313ce9fb4eb)

The information only becomes visual noise due to the way it is presented in.
Realistically, the only elements which should be translucent are the timelines if the right sidebar does eventually become something which you could toggle, but having a toggle for nearly the entirety of the UI in order to not use transparency provides less information and is an extra step you would have to do while mapping to keep showing/hiding information. Additionally, making the playfield smaller in order to account for this is not really a viable solution as the playfield is already smaller than in gameplay / the stable editor, so as I see it there aren't many great solutions which do not involve removing information / shrinking the playfield.

I do not personally think that having those elements be translucent, even if more opaque than the proposed version would clash hard enough with the new designs for it to be a problem and it does make the UI feel a lot more spacious.

**Reply by alexpado** [reply]
What about a toggle/shortcut to switch to a compact UI ? (hiding right sidebar & smaller timeline regarding its height)

This would allow to keep the UI without transparency and still avoid the issue reported in this thread. Mapper could switch the display mode depending on what they want to achieve (map focus / configuration focus).

Edit: Even in compact mode, the whole timeline with the waveform on top of the screen could be hidden as we have the timeline at the bottom

**Reply by confode** [reply]
As well as multiple modes of opacity being a possibility, or even the ability to create UI's that you can save and use for later (obviously limited to the editor).

![image](https://github.com/ppy/osu/assets/121215029/033f54db-b692-4e8f-84ab-7bd811371091) 
![image](https://github.com/ppy/osu/assets/121215029/fc1b4936-d85a-4674-bcf5-d419f007917b)

As you can see here, the lazer editor I feel is less visually appealing when trying to place down objects and such, especially for aspirists. It also seem much more horizontal and closed in due to verticals being pushed down and more being on the screen.

Since the lazer editor is also bigger but with the addition of more features, more transparency than the stable editor I feel would be useful. An opacity bar or a feature similar to one would make the editing experience greater and the feeling of more freedom to move around objects in general.

**Reply by alexpado** [reply]
In the end what you are asking for is making the editor skinnable. As much as I love the idea, the amount of work behind must be huge.

**Comment by peppy** [comment]
I already have similar changes in my editor UI queue, but disagree with what you have for one simple reason: Transparency or hiding of elements should only trigger either:

- When the user requests it
- When required

Still requires exploration as to which option is better.

The opacity levels you've come up with do look pretty good though, visually. I doubt the padding reductions will fly though. Also I think you've inadvertently make the gameplay area larger than actual gameplay (it looks scaled in your screenshots).

**Reply by peppy** [reply]
One other avenue I want to explore is making the playfield easily zoomable (and potentially rotatable). Many image and drawing applications do this, which allows more flexibility and creativity without needing to remove or reduce opacity of the interface.

**Reply by NiceAesth** [reply]
The proposed solutions do sound very promising! I appreciate that this is something that is being looked into.

Initially, I thought of having a UI transparency toggle/submenu in the view menu item (similar to waveform opacity). I think having it as an auto-toggle might be a bit unintuitive, although it would also get the job done.

**Reply by peppy** [reply]
I was hoping it would toggle when the mouse is near or over any of the toolbars (all would become 100% opacity together). An opacity setting could also work, but I'd like to avoid settings if we can with better logic.

**Reply by NiceAesth** [reply]
Ah, I had thought of it the other way around (e.g. making opacity lower when placing objects near the toolbars). The version you proposed does make a lot more sense 😅

**Reply by maotovisk** [reply]
I think the best direction for the opacity is to have the concept of the "open blank canvas": making the backgrounds (at the very least both top and bottom timelines backgrounds) transparent by default, giving the idea of a WAY less claustrophobic editor. I also think that making all backgrounds opaque on hover of those elements is a very interesting direction.

If we consider that 100% UI scaling is the default, the height and padding changes actually make to make the editor playfield feel more like its size in stable. Right now it's basically required to set the UI scaling to 80% to somewhat get close to that size, which by itself it's not a problem but still is a very welcome change.

Some examples on that, using [this osz](https://github.com/ppy/osu/files/12194534/MYTH.ROID.-.VORACITY.osz.zip):
- 1280x960 with 100% UI scaling 

| osu!stable | master | proposed |
| -- | -- | -- |
|![stable reference](https://github.com/ppy/osu/assets/20617307/1a409935-9121-43a6-bf3c-a787fb1bf524)|![master100](https://github.com/ppy/osu/assets/20617307/034cbe21-4bfe-4c5f-be11-81760b74a239)|![changes100](https://github.com/ppy/osu/assets/20617307/3b7a617f-ebf7-4992-982d-cea9e8a32ad5)

- 1280x960 with 80% UI scaling 

| osu!stable | master | proposed |
| -- | -- | -- |
|![stable reference](https://github.com/ppy/osu/assets/20617307/ced4800d-6a9e-4cf0-a3cc-9b4ed37e4b15)|![master80](https://github.com/ppy/osu/assets/20617307/255f82d1-a5c6-4d0c-ac09-bb05cffe4f4d)|![changes80](https://github.com/ppy/osu/assets/20617307/37fb65f2-6a9a-4405-8783-4232d5d3dd8a)

- 1280x960 with 80% UI scaling and BPM indicator disabled:

| osu!stable | master | proposed |
| -- | -- | -- |
|![stable reference](https://github.com/ppy/osu/assets/20617307/ced4800d-6a9e-4cf0-a3cc-9b4ed37e4b15)|![master80_bpm_unticked](https://github.com/ppy/osu/assets/20617307/4e689401-1fca-41b9-9aa2-70a5995792a5)|![changes80_bpm_unticked](https://github.com/ppy/osu/assets/20617307/22251e33-d9ed-4dea-bb94-5faa14b08867)

With those changes, we actually don't need to untick BPM anymore to get somewhat close to the playfield size present in the stable editor, keeping all info showing.

**Comment by bdach** [comment, marked as answer]
https://github.com/ppy/osu/pull/28507

**Reply by NiceAesth** [reply]
thank u i love u
